src/mlpack/methods/reinforcement_learning/sac_impl.hpp

```diff
@@ -0,0 +1,343 @@
+  {
+    arma::mat grad, gradQ, q;
+    arma::colvec singleState = sampledStates.col(i);
+    arma::colvec singlePi;
+    policyNetwork.Forward(singleState, singlePi);
+
+    arma::colvec input = arma::join_vert(singlePi, singleState);
+    learningQ1Network.Forward(input, q);
+    learningQ1Network.Backward(input, -1, gradQ);
+
+    size_t hidden1 = boost::get<mlpack::ann::Linear<> *>
+        (learningQ1Network.Model()[0])->OutputSize();
+    arma::colvec gradQBias = gradQ(input.n_rows * hidden1, 0,
+        arma::size(hidden1, 1));
+    arma::rowvec weightLastLayer = learningQ1Network.Parameters().
+        rows(0, hidden1-1).t();
+    arma::mat gradPolicy = weightLastLayer * gradQBias;
+    policyNetwork.Backward(singleState, gradPolicy, grad);
+    if (i == 0)
+    {
+      gradient.copy_size(grad);
+      gradient.fill(0.0);
+    }
+    gradient += grad;
+  }
+  gradient /= sampledStates.n_cols;
+
+  #if ENS_VERSION_MAJOR == 1
+  policyUpdater.Update(policyNetwork.Parameters(), config.StepSize(), gradient);
+  #else
+  policyNetworkUpdatePolicy->Update(policyNetwork.Parameters(),
+      config.StepSize(), gradient);
+  #endif
+
+  // Update target network
+  if (totalSteps % config.TargetNetworkSyncInterval() == 0)
+    SoftUpdate(0.005);
+}
+
+template <
+  typename EnvironmentType,
+  typename QNetworkType,
+  typename PolicyNetworkType,
+  typename UpdaterType,
+  typename ReplayType
+>
+void SAC<
+  EnvironmentType,
+  QNetworkType,
+  PolicyNetworkType,
+  UpdaterType,
+  ReplayType
+>::SelectAction()
+{
+  // Get the action at current state, from policy.
+  arma::rowvec outputAction;
+  policyNetwork.Predict(state.Encode(), outputAction);
+
+  if (!deterministic)
+  {
+    arma::rowvec noise = arma::randn<arma::rowvec>(outputAction.n_rows) * 0.1;
+    noise = arma::clamp(noise, -0.25, 0.25);
+    outputAction = outputAction + noise;
+  }
+  action.action[0] = outputAction[0];
+}
+
+template <
+  typename EnvironmentType,
+  typename QNetworkType,
+  typename PolicyNetworkType,
+  typename UpdaterType,
+  typename ReplayType
+>
+double SAC<
+  EnvironmentType,
+  QNetworkType,
+  PolicyNetworkType,
+  UpdaterType,
+  ReplayType
+>::Episode()
+{
+  // Get the initial state from environment.
+  state = environment.InitialSample();
+
+  // Track the steps in this episode.
+  size_t steps = 0;
+
+  // Track the return of this episode.
+  double totalReturn = 0.0;
+
+  // Running until get to the terminal state.
+  while (!environment.IsTerminal(state))
+  {
+    if (config.StepLimit() && steps >= config.StepLimit())
+      break;
+    SelectAction();
+
+    // Interact with the environment to advance to next state.
+    StateType nextState;
+    double reward = environment.Sample(state, action, nextState);
+
+    totalReturn += reward;
+    steps++;
+    totalSteps++;
+
+    // Store the transition for replay.
+    replayMethod.Store(state, action, reward, nextState,
+        environment.IsTerminal(nextState), config.Discount());
```

Comment:
In short, just try changing this line to `replayMethod.Store(state, action, reward, nextState, false, config.Discount());` and run the test. You will be surprised to see the agent performing much better!

Reply:
Interesting, such issues are so difficult to find.